### DISCUSSION TITLE
How netbox calculate power draw of a rack?

### DISCUSSION THREAD

**Discussion by older-fox** [netbox-community/netbox#14513] [Help Wanted!] [answered]
I'm trying to get Netbox to calculate the sum of the predefined power consumption of all devices in the current rack, I've tried doing the following list of things：

1. create a "power panel", It represents the power supply control cabinet inside the server room.
2. create a "power feed" based on the "power panel" I just created to a test rack created before. and there's some devices in the rack and predefined power draw.
3. create a "device"  with model "PDU" and 1 "power port", 13 of "power outlets", predefined "Maximum Draw" with 2000Watts, "Allocated Draw" with 1 watts,
4. add this "PDU" to rack,  connect "power port" to "power feed",

So far, the rack preview page successfully displays that the current rack power consumption is equal to the power consumption defined in the PDU, I don't quite understand what happened next. hope someone can point me to the correct approach.

5. Connect other device's power port to PDU's "power outlet"

For example: A device have a 400Watts power port, connect it to PDU's outlet #1, In the rack preview page, nothing changed, seems like it didn't caculate the 400Watts to " rack overall power consumption", and the device preview page of the PDU also shows no change on the power.

**Comment by older-fox** [comment]
Sorry, my english is kinda poor, and huge thanks for everyone who volunteer to help, cheers!

**Comment by candlerb** [comment]
The short answer is it's not well documented. For a long answer (from reverse-engineering the code) see #12837

I believe that rack-level power figures only consider devices which draw from *Power feeds* which are explicitly assigned to that particular rack.

I have found it easier to ignore Netbox's built-in power summarisation, and instead write Custom Reports to do what I want. Actually they are Custom Scripts, because I wanted them to take parameters, and reports don't allow this.

https://github.com/netbox-community/customizations/blob/master/scripts/power_summary.py

**Reply by older-fox** [reply]
Thank you for your answer. I understand that NetBox does not calculate the power consumption of devices below PDU devices for racks. Could you please provide a document/explanation on how I can use this custom script? I have just started learning NetBox and have not yet learned any relevant information about it

**Reply by candlerb** [reply]
Netbox calculates the power consumption *one level* below the PDU devices. e.g. Power Feed -> PDU -> Device is calculated, but Power Feed -> PDU -> PDU -> Device is not.  And if you want to see that utlization under "Rack", then the Power Feed must be assigned to the rack. Please read #12837 for the full details.

Custom scripts are installed via the web UI under Customization > (Reports & Scripts) > Scripts, Add.

For more info, please see the section on [Custom scripts](https://docs.netbox.dev/en/stable/customization/custom-scripts/) in the online documentation.

**Reply by older-fox** [reply]
Thanks, In your description, you mentioned that Power Feed -> PDU -> Device will be calculated into the total power consumption of the rack，But in the tests I did, the PDU was a "device", is this correct? Or is PDU a special thing other than "device"?

The current situation is: I created a PDU of type "Device", added a "Power Port" and a dozen "Power Outlets", then connected the power port to the "Power feed", and then connected the All equipment is connected to these dozen "power outlets". In this case, the rack, including the PDU itself, does not count the power consumption of the equipment connected to the power outlets.

![image](https://github.com/netbox-community/netbox/assets/34577787/6c912bf0-88a6-4cad-a02b-c97c8e86e8f2)
![image](https://github.com/netbox-community/netbox/assets/34577787/e93b8c75-424f-4c2f-bafa-b81b3985d788)
![image](https://github.com/netbox-community/netbox/assets/34577787/ef9f4266-48d0-45de-8978-cd2b2f3fad10)
![image](https://github.com/netbox-community/netbox/assets/34577787/2cc371c8-59e3-49f0-b3ec-7224cb37eeff)

**Reply by candlerb** [reply]
> But in the tests I did, the PDU was a "device", is this correct?

Yes. In the PDU model, you must remember to link each "power outlet" to a "power power" (inlet). This is an attribute of the power outlet, and it models the thru connections. It looks like you've done this though.

Without showing the *full* topology, it's hard to know what you've missed.

Have you tried looking at power utilization at the Power Feed level? If this works, but the Rack power utilization is zero, then this may help narrow down the problem. But the full details of what I did are in #12837

**Reply by older-fox** [reply]
Current topology looks like this
![image](https://github.com/netbox-community/netbox/assets/34577787/97c1878d-3048-4ad7-b451-6ff07648900d)

I can see 1 watt at power feed level, that 1 watt is defind at the PDU's power port section，But nothing changes after I connect devices to it, hope those images could makes you understand more about it.

![image](https://github.com/netbox-community/netbox/assets/34577787/93595d16-1170-4834-b430-00781b228bb0)
![image](https://github.com/netbox-community/netbox/assets/34577787/4b7b45a3-f219-4581-a0e0-b753407f7802)
![image](https://github.com/netbox-community/netbox/assets/34577787/3148e8c5-7a85-40e9-8039-c119a3741cac)
![image](https://github.com/netbox-community/netbox/assets/34577787/c9401d1e-634b-47b0-8e7b-be8553a1a46f)
![image](https://github.com/netbox-community/netbox/assets/34577787/83d03046-f460-40a7-982e-2a223131b903)

**Reply by candlerb** [reply, marked as answer]
> I can see 1 watt at power feed level, that 1 watt is defind at the PDU's power port section

If you define power utilization at the PDU power port level, then it overrides the calculation from the power outlets.

You must leave both the allocated_draw and maximum_draw on the PDU power port empty (null) to enable the summing of usage of devices connected to power outlets.

This is as described in #12837:

> * `get_power_draw` performs a summary calculation on a (PDU) power port
>     * if the power port has either maximum_draw or allocated_draw statically assigned, then it returns both these values (and if one of these values is null then it is converted to 0)
>     * otherwise, it finds all the ports from `get_downstream_powerports` and sums the maximum_draw and allocated_draw from these
